libs/@blockprotocol/graph/src/stdlib/subgraph/element/entity.ts

```diff
@@ -54,80 +66,82 @@ export const getEntity = (
+      .map((revisionTimestamp) => {
+        const vertex = mustBeDefined(entityRevisions[revisionTimestamp]);
 
-      // Last element in the array (latest version), so we assume an half-closed interval (unbounded on the upper-bound)
-      if (nextRevisionVersion === undefined) {
-        if (revisionVersion <= targetTime) {
-          if (isEntityVertex(entityRevisions[revisionVersion]!)) {
-            targetVersion = revisionVersion;
-          }
-        }
-        break;
-      } else if (
-        revisionVersion <= targetTime &&
-        targetTime < nextRevisionVersion
-      ) {
-        if (isEntityVertex(entityRevisions[revisionVersion]!)) {
-          targetVersion = revisionVersion;
-          break;
+        if (!isEntityVertex(vertex)) {
+          throw new Error(
+            `Found non-entity vertex associated with EntityId: ${entityId}`,
+          );
         }
-      }
-    }
-
-    entityRevision =
-      targetVersion !== undefined
-        ? entityRevisions[targetVersion]!.inner
-        : undefined;
+        return vertex.inner;
+      })
+      .find((entity) => {
+        return intervalContainsTimestamp(
+          entity.metadata.temporalVersioning[
+            subgraph.temporalAxes.resolved.variable.axis
+          ],
+          targetTime,
+        );
+      });
   }
-
-  return entityRevision;
 };
 
 /**
- * Returns all `Entity`s within the vertices of the subgraph that match a given `EntityId`
+ * Returns all {@link Entity} revisions within the vertices of the subgraph that match a given {@link EntityId}.
  *
- * @param subgraph
- * @param entityId
+ * When querying a subgraph with support for temporal versioning, it optionally constrains the search to a given
+ * {@link NonNullTimeInterval}.
+ *
+ * @param {Subgraph }subgraph
+ * @param {EntityId} entityId
+ * @param {NonNullTimeInterval} [interval]
  */
-export const getEntityRevisionsByEntityId = (
-  subgraph: Subgraph,
+export const getEntityRevisionsByEntityId = <Temporal extends boolean>(
+  subgraph: Subgraph<Temporal>,
   entityId: EntityId,
-): Entity[] => {
+  interval?: Temporal extends true ? NonNullTimeInterval : never,
+): Entity<Temporal>[] => {
   const versionObject = subgraph.vertices[entityId];
 
   if (!versionObject) {
     return [];
   }
-  const entityVertices = Object.values(versionObject);
 
-  return entityVertices.filter(isEntityVertex).map((vertex) => {
-    return vertex.inner;
-  });
+  if (interval !== undefined) {
+    return typedEntries(versionObject)
+      .filter(([startTime, _entityVertex]) =>
+        intervalContainsTimestamp(interval, startTime),
+      )
+      .map(([_startTime, entityVertex]) => entityVertex)
+      .filter(isEntityVertex)
+      .map((entityVertex) => entityVertex.inner)
+      .filter((entity) =>
+        intervalContainsInterval(
+          interval,
+          entity.metadata.temporalVersioning[
+            subgraph.temporalAxes.resolved.variable.axis
+          ],
+        ),
```

Comment:
Hm, interesting (and great catch). I think we do want `intervalOverlapsInterval` instead of `Contains` here. We want to return any entity which was active over that range, the fact that it was still active after the search range doesn't matter (and would be filtered out by this).

I'll update and add a comment explaining this

Reply:
Have pushed

Reply:
Thanks! :)